### ISSUE TITLE
Impossible to decorate Page fields for multiple drivers

### ISSUE THREAD

**Issue by FiMka** [appium/java-client#221]
Hello!
What is described below isn't an issue (it's by design), but could you possibly suggest something to solve my issue.
My tests use @AndroidFindBy annotations to inject WebElements. My test cases involve multiple devices, so I use [selenium grid approach](http://stackoverflow.com/questions/18719755/multi-devices-support-in-android). This implies creating of multiple AndroidDriver instances.
Could you suggest how to decorate Page objects when using several drivers?
```java
PageFactory.initElements(new AppiumFieldDecorator(driver,
    MAX_WAIT_FOR_ELEMENT_SEC, TimeUnit.SECONDS), bean);
```

**Comment by TikhomirovSergey** [comment]
Hi @FiMka 

I'm trying to fugure out your problem. 
Am I correcly understand that you create a several drivers at one test session (e.g. one test method opens few Appium sessions and checks your app on few devices/emulators at an one time)? If it is true why is it not the parallel testing?

Ok. If you have to use several drivers at one test method then you shoul create an instance of  AppiumFieldDecorator for each one driver.

```java
driver1 = new AndroidDriver<>(constructor parameters) ;
driver2 = new AndroidDriver<>(constructor parameters) ;
//

AppiumFieldDecorator afd1 = new AppiumFieldDecorator(driver,
    MAX_WAIT_FOR_ELEMENT_SEC, TimeUnit.SECONDS);

AppiumFieldDecorator afd2 = new AppiumFieldDecorator(driver,
    MAX_WAIT_FOR_ELEMENT_SEC, TimeUnit.SECONDS);

//your code
//lots of code :)
//

PageFactory.initElements(afd1, bean1);
PageFactory.initElements(afd2, bean2);

//or if your page object is the same for all drivers
//
PageFactory.initElements(afd1, bean);
PageFactory.initElements(afd2, bean);
```

If you have the parallel testing or you are able to organize it (the same test method is being run with different parameters in different threads at an one time). The code sample above becomes much simplified.

**Comment by TikhomirovSergey** [comment]
It would be cool is there was a method like that on Selenium side:

```java
PageFactory.initElements(searchContext, decoratorInstance, pageObjectInstance);
```

PageFactory has methods that you can if you'll go to link. [Selenium PageFactory](https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/support/PageFactory.java)

**Comment by FiMka** [comment]
Hi Sergey! I do really appreciate your assistance in solving my problem!

> **As Sergey Tikhomirov said**: Am I correcly understand that you create a several drivers at one test session (e.g. one test method opens few Appium sessions and checks your app on few devices/emulators at an one time)? If it is true why is it not the parallel testing?

You understand this absolutely right. I can give a simple example when this can be needed: [Two devices are chatting with each other](https://www.youtube.com/watch?v=5UlmnplfCRc). This use case is real close to what I have. With parallel testing the synchronization between devices may be more complicated.

> **As Sergey Tikhomirov said**: Ok. If you have to use several drivers at one test method then you should create an instance of AppiumFieldDecorator for each one driver.

Thank you! I'll definitely try this out.

> **As Sergey Tikhomirov said**: PageFactory has methods that you can if you'll go to link. Selenium PageFactory

Got it, thanks. This is what I use right now.

**Comment by FiMka** [comment]
> **As Sergey Tikhomirov said**: Ok. If you have to use several drivers at one test method then you should create an instance of AppiumFieldDecorator for each one driver.

I've just remembered that I tried this already as follows:
```java
  @Override
  public Object postProcessBeforeInitialization(final Object bean, final String beanName)
      throws BeansException {
    if (bean.getClass().isAnnotationPresent(Page.class)) {
      for (final AndroidDriver driver : drivers) {
        PageFactory.initElements(new AppiumFieldDecorator(driver, MAX_WAIT_FOR_ELEMENT_SEC,
          TimeUnit.SECONDS), bean);
      }
    }
    return bean; // return decorated page
  }
```
I use Spring in my project, all Page objects (represented as Spring beans) are decorated in postProcessBeforeInitialization (in a class implementing [BeanPostProcessor](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/beans/factory/config/BeanPostProcessor.html)). So each page bean is decorated only with latest driver from the list. I hope I'm explaining clear.

It seems I have to throw out Spring from my project or forget about @AndroidFindBy annotations.

**Comment by TikhomirovSergey** [comment]
It seems that you are doing something complicated. I don't think that there it is needed to throw out Spring or Appium annotations... Why is it not possible to decorate a created bean?

**Comment by TikhomirovSergey** [comment]
Ok. If it is needed to use this instantion then postprocessing should know what driver is "free". If it possible to use only the last element of the driver collection?

**Comment by FiMka** [comment]
Sergey, let me rephrase my question. If I have a page like:
```java
public class SomePage {

  @AndroidFindBy(id = PACKAGE + ":id/someButton”)
  private WebElement someButton;

}
```
And this page is decorated for each one driver (like you suggested). How to "switch" between drivers (devices) during testing? Because when using @AndroidFindBy annotations I don't operate anymore with driver directly in the tests. I use this once only - when initializing (injecting) page fields (web elements).

**Comment by TikhomirovSergey** [comment]
Ah. Now your question is clear, I think...

So when you create a driver 

```java
driver1 = new AndroidDriver<>(URL of a machine where Appium server 1 is launced, capabilities) ;
driver2 = new AndroidDriver<>(URL of a machine where Appium server 2 is launced, capabilities) ;
```

two different sessions are being created. Each one session is related to the certain server. Just because appium server doesn't works with two or more devices/emulators so each one session is linked with the certain device/emulator.

When you are performing the following 

```java
AppiumFieldDecorator afd1 = new AppiumFieldDecorator(driver1,
    MAX_WAIT_FOR_ELEMENT_SEC, TimeUnit.SECONDS);

AppiumFieldDecorator afd2 = new AppiumFieldDecorator(driver2,
    MAX_WAIT_FOR_ELEMENT_SEC, TimeUnit.SECONDS);

//your code
//lots of code :)
//

PageFactory.initElements(afd1, bean1);
PageFactory.initElements(afd2, bean2);
```

you are populating fields of two different instances (bean1, bean2) using two different drivers (driver1 and driver2). 

Ok.

```java
public class SomePage {

  @AndroidFindBy(id = PACKAGE + ":id/someButton”)
  private WebElement someButton; //it is not a real element. It is proxy object created by CGLib.
  //It wraps the certain search context. When you are trying to do something with element 
  //then:
  //- it performs the searching
  // - invoces the method of the real element that has been found

  public void clickSomeButton(){
      someButton.click(); //here the real element is being found and click is being berformed
  } 

}
```

So. There is nothing to do in order to switch between devices. 

 ```java
bean1.clickSomeButton(); //perfoms the action on device 1/ emilator 1
bean2.clickSomeButton(); //perfoms the action on device 2/ emilator 2
```

I think it is all. If there are more questions you are free to ask :)

**Comment by FiMka** [comment]
Sergey, thanks for the clarification! This is how I thought it should work.
> **As Sergey Tikhomirov said**: So. There is nothing to do something special in order to switch between devices.
```java
bean1.clickSomeButton(); //perfoms the action on device 1/ emilator 1
bean2.clickSomeButton(); //perfoms the action on device 2/ emilator 2
```

I use [Spring](https://peterkedemo.wordpress.com/2013/03/30/writing-good-selenium-tests-with-page-objects-and-spring/) for pages, so they look like:

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Component
public @interface Page {
}
...
@Page
public class SomePage {

  @AndroidFindBy(id = PACKAGE + ":id/someButton”)
  private WebElement someButton;

  @When("Press some button")
  public void pressSomeButton() {
    someButton.click();
  }

  @Then("Button is pressed")
  public void verifyButtonIsPressed() {
    Assert.assertTrue(...);
  }

}
```

I use [JBehave](http://mkuthan.github.io/blog/2014/05/29/acceptance-testing-using-jbehave-spring-framework-and-maven/) for tests, so they look like:

```
Scenario: This test case presses some button and verifies it is pressed.

When Press some button
Then Button is pressed
```

JBehave knows about pages because of this:

```java
public abstract class AbstractSpringJBehaveStory extends JUnitStory {
  ...
  @Override
  public InjectableStepsFactory stepsFactory() {
    return new SpringStepsFactory(configuration(), applicationContext);
  ...
}
```

So the workflow can be described as follows:

1. JBehave scenario is executed with SpringJUnit4ClassRunner
2. Spring starts scanning all beans
3. Spring detects Page beans (annotated with @Page) and creates **ONLY ONE** bean for each page
4. JBehave script is executed step by step, Page beans are detected automatically in the application context

So in this scheme there is no many options on how to switch between a same page decorated with different drivers. Only a single instance of each page is on the context. This can be "fixed" if, for example, forget about @AndroidFindBy:

```java
public class SomePage {

  private By someButton = By.id(PACKAGE + ":id/someButton”));

  @When("Press some button")
  public void pressSomeButton() {
    driver.findElement(someButton).click(); // driver is used directly and be switched
  }

  @Then("Button is pressed")
  public void verifyButtonIsPressed() {
    Assert.assertTrue(...);
  }

}
```

**Comment by TikhomirovSergey** [comment]
There is no conversations anymore.
It is defenitly the problem of user's framework or used third party lidrary.
Maybe the fix of #228 will simplify this case.
